rust_src/src/buffers.rs

```diff
@@ -1228,4 +1230,67 @@ pub extern "C" fn rust_syms_of_buffer() {
     defvar_per_buffer!(header_line_format_, "header-line-format", Qnil);
 }
 
+/// Change current buffer's name to NEWNAME (a string).  If second arg
+/// UNIQUE is nil or omitted, it is an error if a buffer named NEWNAME
+/// already exists.  If UNIQUE is non-nil, come up with a new name
+/// using `generate-new-buffer-name'.  Interactively, you can set
+/// UNIQUE with a prefix argument.  We return the name we actually
+/// gave the buffer.  This does not change the name of the visited
+/// file (if any).
+#[lisp_fn(
+    min = "1",
+    intspec = "(list (read-string \"Rename buffer (to new name): \" nil 'buffer-name-history (buffer-name (current-buffer))) current-prefix-arg)"
+)]
+pub fn rename_buffer(newname: LispStringRef, unique: LispObject) -> LispStringRef {
+    if newname.is_empty() {
+        error!("Empty string is invalid as a buffer name")
+    }
+
+    let mut current_buffer = ThreadState::current_buffer_unchecked();
+
+    let newname = match get_buffer(LispBufferOrName::Name(newname.into())) {
+        None => newname,
+        Some(tem) => {
+            // Don't short-circuit if UNIQUE is t.  That is a useful
+            // way to rename the buffer automatically so you can
+            // create another with the original name.  It makes UNIQUE
+            // equivalent to
+            // (rename-buffer (generate-new-buffer-name NEWNAME)).
+            if unique.is_nil() && (tem == current_buffer) {
+                return current_buffer.name_.into();
+            }
+            if unique.is_not_nil() {
+                generate_new_buffer_name(newname, current_buffer.name_)
+            } else {
+                error!("Buffer name `{}' is in use", newname)
+            }
```

Comment:
This can be restructured right?

```
           if unique.is_nil() {
               if tem == current_buffer {
                   return current_buffer.name_.into();
               }
               error!("Buffer name `{}' is in use", newname);
           } else { 
               generate_new_buffer_name(newname, current_buffer.name_)
           }
```

I think this version reads better.

```
           if unique.is_not_nil() {
               generate_new_buffer_name(newname, current_buffer.name_)
           } else { 
               if tem == current_buffer {
                   return current_buffer.name_.into();
               }
               error!("Buffer name `{}' is in use", newname);
           }
```

Reply:
/Technically/ this version has a little more overhead since we always generate a new buffer name in the common case. Think I'll quickly fix that as well.